### ISSUE TITLE
The from_pretrained method of the DiffusionPipeline doesn't set the passed modules' mode to evaluation.

### ISSUE THREAD

**Issue by wtyuan96** [huggingface/diffusers#9660]
### Describe the bug

When passing some modules to the from_pretrained method of DiffusionPipeline,  this method will instantizate the pipeline using passed modules instead of loading from the cached folder (see [here](https://github.com/huggingface/diffusers/blob/6a5f06488c0d88c1827c016835cd5f64abe4b52c/src/diffusers/pipelines/pipeline_utils.py#L885) and [here](https://github.com/huggingface/diffusers/blob/6a5f06488c0d88c1827c016835cd5f64abe4b52c/src/diffusers/pipelines/pipeline_utils.py#L932)).

However, **passed modules** in training mode was not set to evaluation mode using the `eval()` function, resulting in the inference processes of many example training scripts (for example [tran_controlnet.py](https://github.com/huggingface/diffusers/blob/main/examples/controlnet/train_controlnet.py) and [train_text_to_image.py)](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py) actually being flawed.

### Reproduction

This issue can be reproduced by setting a breakpoint [here](https://github.com/huggingface/diffusers/blob/6a5f06488c0d88c1827c016835cd5f64abe4b52c/examples/controlnet/train_controlnet.py#L1057) and then printing the training attribute of controlnet:
```python
print(controlnet.training) #  The output is True
down_block_res_samples, mid_block_res_sample = controlnet(
    noisy_latents,
    timesteps,
    encoder_hidden_states=encoder_hidden_states,
    controlnet_cond=controlnet_image,
    return_dict=False,
)
```

### System Info

- 🤗 Diffusers version: 0.31.0.dev0
- Platform: Linux-5.10.134
- Running on Google Colab?: No
- Python version: 3.10.14
- PyTorch version (GPU?): 2.4.0+cu118 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Huggingface_hub version: 0.24.6
- Transformers version: 4.44.2
- Accelerate version: 0.33.0
- PEFT version: 0.13.1
- Bitsandbytes version: not installed
- Safetensors version: 0.4.4
- xFormers version: not installed

### Who can help?

@sayakpaul @DN6

**Comment by a-r-r-o-w** [comment]
Hey. Yes, this seems like incorrect behaviour, I've verified this and you're right - thanks for reporting!

We only set the model to eval mode during modeling `from_pretrained` [here](https://github.com/huggingface/diffusers/blob/6a5f06488c0d88c1827c016835cd5f64abe4b52c/src/diffusers/models/modeling_utils.py#L856). There should be a call to `.eval()` before validation. For testing, we reload all modules, so this wouldn't be required there. Would you be so kind to open PRs fixing this?

**Comment by wtyuan96** [comment]
> Hey. Yes, this seems like incorrect behaviour, I've verified this and you're right - thanks for reporting!
> 
> We only set the model to eval mode during modeling `from_pretrained` [here](https://github.com/huggingface/diffusers/blob/6a5f06488c0d88c1827c016835cd5f64abe4b52c/src/diffusers/models/modeling_utils.py#L856). There should be a call to `.eval()` before validation. For testing, we reload all modules, so this wouldn't be required there. Would you be so kind to open PRs fixing this?

At first, I intended to directly submit a pull request rather than an issue. However, I found that fixing this bug is not as straightforward, so let me elaborate.

The most elegant way would be to set the mode of the passed module to "evaluation" within the `from_pretrained` method of `DiffusionPipeline`, similar to how the reloaded module is handled. The advantage of this approach is that it doesn't require modifying the example training scripts. However, after the pipeline inference is completed, the passed module needs to be reset to "training" mode, which is not easy to implement.

Another approach is to add `eval()` and `train()` methods around the pipeline in the `log_validation` method of the training scripts. This way, however, would require modifying many of the example training scripts.

Which of these approaches do you think is more reasonable?

**Comment by sayakpaul** [comment]
Thanks for bringing this to our attention! Appreciate the details. 

I think we can set `eval()` and `train()` manually from the training scripts to make things explicit (which aligns with our philosophy of the library). I won't prefer touching the pipeline code here TBH. 

The reason why we probably didn't notice any bugs in the generation quality is because coincidentally, the train and eval modes of these models don't differ much in our training scripts. For example, we know that BatchNorm, Dropout, etc. have different train and eval mode behaviors. But since we don't train with these things, we don't see any effects in the results.
